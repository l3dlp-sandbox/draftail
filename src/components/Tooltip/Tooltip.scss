@import "~tippy.js/dist/tippy.css";

.#{$DRAFTAIL}Tooltip {
  background: none;
  color: inherit;
  border-radius: 0;
  font-size: inherit;

  .tippy-content {
    padding: 0;
  }
}

.#{$DRAFTAIL}Tooltip__target {
  position: absolute;
  pointer-events: none;
  visibility: hidden;
}

.#{$DRAFTAIL}Tooltip__backdrop {
  position: relative;
  top: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0);
  transition: background-color $draftail-toolbar-tooltip-duration 0s;
  z-index: $draftail-page-overlay-z-index;
  pointer-events: none;
}

.#{$DRAFTAIL}Tooltip__backdrop--visible {
  position: fixed;
  background-color: rgba(0, 0, 0, 0.1);
}
