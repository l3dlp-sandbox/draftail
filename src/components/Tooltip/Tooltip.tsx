import React, { useEffect, useRef, useState } from "react";
import Tippy, { TippyProps } from "@tippyjs/react";

const hideTooltipOnEsc = {
  name: "hideOnEsc",
  defaultValue: true,
  fn({ hide }: { hide: () => void }) {
    function onKeyDown(e: KeyboardEvent) {
      if (e.key === "Escape") {
        hide();
      }
    }

    return {
      onShow() {
        document.addEventListener("keydown", onKeyDown);
      },
      onHide() {
        document.removeEventListener("keydown", onKeyDown);
      },
    };
  },
};

const zeroWidthSpacer = "\u200B";

const tippyPlugins = [hideTooltipOnEsc];

// https://atomiks.github.io/tippyjs/v6/all-props/#duration
const duration: [number, number] = [300, 0];

export interface TooltipPosition {
  top: number;
  left: number | string;
}

export type TooltipPlacement = TippyProps["placement"];
export type PopperInstance = Parameters<NonNullable<TippyProps["onMount"]>>[0];

export interface TooltipProps {
  shouldOpen: boolean;
  getTargetPosition?: (editorRect: DOMRect) => TooltipPosition | null;
  content: React.ReactNode;
  children?: React.ReactElement;
  showBackdrop?: boolean;
  zIndex?: number;
  placement?: TooltipPlacement;
  onHide?: () => void;
  onMount?: (instance: PopperInstance) => void;
}

const Tooltip = ({
  content,
  children,
  shouldOpen,
  getTargetPosition,
  showBackdrop = false,
  zIndex = 100,
  placement = "top" as TooltipPlacement,
  onHide,
  onMount,
}: TooltipProps) => {
  const parentRef = useRef<HTMLDivElement>(null);
  const [position, setPosition] = useState<TooltipPosition | null>();

  useEffect(() => {
    if (!children && getTargetPosition && parentRef.current) {
      if (shouldOpen) {
        const editor = parentRef.current.closest<HTMLDivElement>(
          "[data-draftail-editor]",
        );
        const editorRect = editor!.getBoundingClientRect();
        setPosition(getTargetPosition(editorRect));
      } else {
        setPosition(null);
      }
    }
  }, [shouldOpen, getTargetPosition, children]);

  const visible = shouldOpen && Boolean(position || children);

  return (
    <>
      {showBackdrop ? (
        <div
          className={`Draftail-Tooltip__backdrop${
            shouldOpen ? " Draftail-Tooltip__backdrop--visible" : ""
          }`}
        />
      ) : null}
      <div
        hidden
        contentEditable="false"
        suppressContentEditableWarning
        ref={parentRef}
      />
      <Tippy
        className="Draftail-Tooltip"
        visible={visible}
        interactive
        onHide={onHide}
        onClickOutside={onHide}
        onMount={onMount}
        placement={placement}
        maxWidth="100%"
        zIndex={zIndex}
        duration={duration}
        arrow={false}
        appendTo={() => {
          const editor = parentRef.current!.closest<HTMLDivElement>(
            "[data-draftail-editor]",
          );
          const tooltipParent = editor!.querySelector(
            "[data-draftail-tooltip-parent]",
          );
          return tooltipParent as HTMLDivElement;
        }}
        plugins={tippyPlugins}
        content={content}
      >
        {children || (
          <div
            className="Draftail-Tooltip__target"
            style={position || undefined}
          >
            {zeroWidthSpacer}
          </div>
        )}
      </Tippy>
    </>
  );
};

export default Tooltip;
