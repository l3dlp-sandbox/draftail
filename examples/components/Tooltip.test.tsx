import React from "react";
import { shallow } from "enzyme";

import Tooltip from "./Tooltip";

const target = {
  top: 1,
  left: 1,
  width: 12,
  height: 1200,
};

describe("Tooltip", () => {
  it("#direction top", () => {
    expect(
      shallow(
        <Tooltip target={target} direction="top" textDirectionality="LTR">
          Test
        </Tooltip>,
      ),
    ).toMatchInlineSnapshot(`
      <div
        className="Tooltip Tooltip--top"
        role="tooltip"
        style={
          Object {
            "insetInlineStart": 7,
            "top": 1201,
          }
        }
      >
        Test
      </div>
    `);
  });

  it("#direction start", () => {
    expect(
      shallow(
        <Tooltip target={target} direction="start" textDirectionality="LTR">
          Test
        </Tooltip>,
      ),
    ).toMatchInlineSnapshot(`
      <div
        className="Tooltip Tooltip--start"
        role="tooltip"
        style={
          Object {
            "insetInlineStart": 13,
            "top": 601,
          }
        }
      >
        Test
      </div>
    `);
  });

  it("#direction top-start", () => {
    expect(
      shallow(
        <Tooltip target={target} direction="top-start" textDirectionality="LTR">
          Test
        </Tooltip>,
      ),
    ).toMatchInlineSnapshot(`
      <div
        className="Tooltip Tooltip--top-start"
        role="tooltip"
        style={
          Object {
            "insetInlineStart": 1,
            "top": 1201,
          }
        }
      >
        Test
      </div>
    `);
  });
});
