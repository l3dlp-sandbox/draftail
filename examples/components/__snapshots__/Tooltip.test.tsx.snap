// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip #direction start 1`] = `
<div
  className="Tooltip Tooltip--start"
  role="tooltip"
  style={
    Object {
      "insetInlineStart": 13,
      "top": 601,
    }
  }
>
  Test
</div>
`;

exports[`Tooltip #direction top 1`] = `
<div
  className="Tooltip Tooltip--top"
  role="tooltip"
  style={
    Object {
      "insetInlineStart": 7,
      "top": 1201,
    }
  }
>
  Test
</div>
`;

exports[`Tooltip #direction top-start 1`] = `
<div
  className="Tooltip Tooltip--top-start"
  role="tooltip"
  style={
    Object {
      "insetInlineStart": 1,
      "top": 1201,
    }
  }
>
  Test
</div>
`;
